File: TotalCrossVM/src/nm/cielo/printer/PrinterManager.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda. 
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

#include "tcvm.h"

#if defined(ANDROID)
#include "android/PrinterManager_c.h"
#endif

//////////////////////////////////////////////////////////////////////////
TC_API void tcspPM_internalPrintText_ssp(NMParams p) // totalcross/cielo/sdk/printer/PrinterManager private native void internalPrintText(String textToPrint, String printerAttributes, totalcross.cielo.printer.PrinterManager.PrinterListenerInternal printerListener);
{
    TCObject printerManager = p->obj[0];
    TCObject textToPrint = p->obj[1];
    TCObject printerAttributes = p->obj[2];
    TCObject printerListener = p->obj[3];
    
#if defined (ANDROID)
    // first, lock the printerListener to make sure it won't be collected by the GC
    if (printerListener != null) {
        setObjectLock(printerListener, LOCKED);
    }
    // make native android call
    cieloPrintManagerPrintText(textToPrint, printerAttributes, printerListener);
#endif
}
#ifdef ENABLE_TEST_SUITE
//#include "PrinterManager_test.h"
#endif

File: TotalCrossSDK/src/main/java/totalcross/cielo/sdk/printer/PrinterManager.java
// Copyright (C) 2019-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

package totalcross.cielo.sdk.printer;

import java.util.Map;

import com.totalcross.annotations.ReplacedByNativeOnDeploy;

import totalcross.cielo.sdk.order.PrinterListener;
import totalcross.sys.Vm;

/**
 * Usage example:
 * 
 * <pre>
 * HashMap<String, Integer> printerAttributes = new HashMap<>();
 * 
 * printerAttributes.put(PrinterAttributes.KEY_ALIGN, PrinterAttributes.VAL_ALIGN_CENTER);
 * printerAttributes.put(PrinterAttributes.KEY_TYPEFACE, 1);
 * printerAttributes.put(PrinterAttributes.KEY_TEXT_SIZE, 20);
 * 
 * String textToPrint = "TEXT TO PRINT";
 * PrinterManager.getInstance()
 *               .printText(textToPrint, printerAttributes);
 * </pre>
 */
public class PrinterManager {

    private static PrinterManager instance;

    private PrinterManager() {
    }

    public static PrinterManager getInstance() {
        if (instance == null) {
            instance = new PrinterManager();
        }
        return instance;
    }

    static class PrinterListenerInternal {
        PrinterListener printerListener;

        PrinterListenerInternal(PrinterListener printerListener) {
            this.printerListener = printerListener;
        }
    }

    public void printText(String textToPrint, Map<String, Integer> printerAttributes) {
        printText(textToPrint, printerAttributes, null);
    }

    private void printText(String textToPrint, Map<String, Integer> printerAttributes,
            PrinterListener printerListener) {
        StringBuffer sb = new StringBuffer();
        for (Map.Entry<String, Integer> entry : printerAttributes.entrySet()) {
            sb.append(entry.getKey());
            sb.append('=');
            sb.append(entry.getValue());
            sb.append('&');
        }
        sb.deleteCharAt(sb.length() - 1);

        internalPrintText(
                textToPrint,
                sb.toString(),
                printerListener == null ? null : new PrinterListenerInternal(printerListener));
    }

    @ReplacedByNativeOnDeploy
    private void internalPrintText(String textToPrint, String printerAttributes,
            PrinterListenerInternal printerListener) {
        Vm.debug("PRINT ONLY AVAILABLE ON CIELO DEVICE");
    }
}
